WcaOnRails/app/webpacker/javascript/hooks/useNestedInputUpdater.js

```diff
@@ -0,0 +1,29 @@
+import { useCallback } from 'react';
+import _ from 'lodash';
+
+// This aims to provide a quick wrapper to:
+//   - create an updater for a given path in a state created by useState or similar.
+//   - take into account how react-semantic-ui provides the input value through its
+//   onChange method, which must have this signature:
+//   onChange(event: ChangeEvent, data: object)
+// Example usage:
+// const setMyProp = useNestedInputUpdater(setGlobalState, 'myProp');
+//
+// It can be used directly like any other set method:
+// setMyProp('my val');
+// Or passed as the 'onChange' callback of a react-semantic-ui input.
+
+const useNestedInputUpdater = (updater, path) => useCallback((ev, data = undefined) => {
+  const value = data ? data.value : ev;
+  // updater is assumed to come from useState or similar, so we can pass it a
+  // function that act based on the previous state.
+  updater((prevState) => {
+    // This is a react state, we can't just modify something in prevValue,
+    // we need to create a brand new copy.
+    const newState = { ...prevState };
+    _.set(newState, path, value);
```

Comment:
`{ ...prevState }` only does a shallow copy, so if the user passes a deep path we would still mutate some of the underlying state.

Reply:
As far as react is concerned what matters is that the returned state's ref is different though, right?
It looks to me that react detects that the returned state is different and does trigger a component update.
At least it seems to work fine when changing an element of a nested array [here](https://github.com/thewca/worldcubeassociation.org/pull/6360/files#diff-872fe3eb10f149f9aa4c589d8b81095773c20b2e67e484ccb5242bd9bf90cc5cR27).

But maybe it's just lucky / undefined behavior and we do need a deep clone?